README.md
textblob-de README
==================

[![textblob\_de - latest PyPI version](https://img.shields.io/pypi/v/textblob-de.svg)](https://pypi.python.org/pypi/textblob-de/)
[![Build Status](https://travis-ci.com/markuskiller/textblob-de.svg?branch=dev)](https://travis-ci.com/markuskiller/textblob-de)
[![Documentation Status](https://readthedocs.org/projects/textblob-de/badge/?version=latest)](https://textblob-de.readthedocs.io/en/latest/?badge=latest)
[![Number of PyPI downloads](https://img.shields.io/pypi/dm/textblob-de.svg)](https://pypi.python.org/pypi/textblob-de/)
[![LICENSE info](https://img.shields.io/github/license/markuskiller/textblob-de.svg)](http://choosealicense.com/licenses/mit/)

German language support for
[TextBlob](http://textblob.readthedocs.org/en/dev/) by Steven Loria.

This python package is being developed as a `TextBlob` **Language
Extension**. See [Extension
Guidelines](https://textblob.readthedocs.org/en/dev/contributing.html)
for details.

Features
--------

-   All directly accessible `textblob_de` classes (e.g. `Sentence()` or
    `Word()`) are initialized with default models for German
-   Properties or methods that do not yet work for German raise a
    `NotImplementedError`
-   German sentence boundary detection and tokenization
    (`NLTKPunktTokenizer`)
-   Consistent use of specified tokenizer for all tools
    (`NLTKPunktTokenizer` or `PatternTokenizer`)
-   Part-of-speech tagging (`PatternTagger`) with keyword
    `include_punc=True` (defaults to `False`)
-   Tagset conversion in `PatternTagger` with keyword
    `tagset='penn'|'universal'|'stts'` (defaults to `penn`)
-   Parsing (`PatternParser`) with all `pattern` keywords, plus
    `pprint=True` (defaults to `False`)
-   Noun Phrase Extraction (`PatternParserNPExtractor`)
-   Lemmatization (`PatternParserLemmatizer`)
-   Polarity detection (`PatternAnalyzer`) - Still **EXPERIMENTAL**,
    does not yet have information on subjectivity
-   Full `pattern.text.de` API support on Python3
-   See [working features
    overview](http://langui.ch/nlp/python/textblob-de-dev/) for details

Installing/Upgrading
--------------------

    $ pip install -U textblob-de
    $ python -m textblob.download_corpora

Or the latest development release (apparently this does not always work
on Windows see [issues \#1744/5](https://github.com/pypa/pip/pull/1745)
for details):

    $ pip install -U git+https://github.com/markuskiller/textblob-de.git@dev
    $ python -m textblob.download_corpora

Note

`TextBlob` will be installed/upgraded automatically when running
`pip install`. The second line (`python -m textblob.download_corpora`)
downloads/updates nltk corpora and language models used in `TextBlob`.

Usage
-----

``` python
>>> from textblob_de import TextBlobDE as TextBlob
>>> text = '''Heute ist der 3. Mai 2014 und Dr. Meier feiert seinen 43. Geburtstag. 
Ich muss unbedingt daran denken, Mehl, usw. für einen Kuchen einzukaufen. Aber leider 
habe ich nur noch EUR 3.50 in meiner Brieftasche.'''
>>> blob = TextBlob(text)
>>> blob.sentences
[Sentence("Heute ist der 3. Mai 2014 und Dr. Meier feiert seinen 43. Geburtstag."),
 Sentence("Ich muss unbedingt daran denken, Mehl, usw. für einen Kuchen einzukaufen."),
 Sentence("Aber leider habe ich nur noch EUR 3.50 in meiner Brieftasche.")]
>>> blob.tokens
WordList(['Heute', 'ist', 'der', '3.', 'Mai', ...]
>>> blob.tags
[('Heute', 'RB'), ('ist', 'VB'), ('der', 'DT'), ('3.', 'LS'), ('Mai', 'NN'), 
('2014', 'CD'), ...]
# Default: Only noun_phrases that consist of two or more meaningful parts are displayed.
# Not perfect, but a start (relies heavily on parser accuracy)
>>> blob.noun_phrases
WordList(['Mai 2014', 'Dr. Meier', 'seinen 43. Geburtstag', 'Kuchen einzukaufen', 
'meiner Brieftasche'])
```

``` python
>>> blob = TextBlob("Das Auto ist sehr schön.")
>>> blob.parse()
'Das/DT/B-NP/O Auto/NN/I-NP/O ist/VB/B-VP/O sehr/RB/B-ADJP/O schön/JJ/I-ADJP/O'
>>> from textblob_de import PatternParser
>>> blob = TextBlobDE("Das ist ein schönes Auto.", parser=PatternParser(pprint=True, lemmata=True))
>>> blob.parse()
      WORD   TAG    CHUNK   ROLE   ID     PNP    LEMMA   

       Das   DT     -       -      -      -      das     
       ist   VB     VP      -      -      -      sein    
       ein   DT     NP      -      -      -      ein     
   schönes   JJ     NP ^    -      -      -      schön   
      Auto   NN     NP ^    -      -      -      auto    
         .   .      -       -      -      -      .       
>>> from textblob_de import PatternTagger
>>> blob = TextBlob(text, pos_tagger=PatternTagger(include_punc=True))
[('Das', 'DT'), ('Auto', 'NN'), ('ist', 'VB'), ('sehr', 'RB'), ('schön', 'JJ'), ('.', '.')]
```

``` python
>>> blob = TextBlob("Das Auto ist sehr schön.")
>>> blob.sentiment
Sentiment(polarity=1.0, subjectivity=0.0)
>>> blob = TextBlob("Das ist ein hässliches Auto.")     
>>> blob.sentiment
Sentiment(polarity=-1.0, subjectivity=0.0)
```

Warning

**WORK IN PROGRESS:** The German polarity lexicon contains only
uninflected forms and there are no subjectivity scores yet. As of
version 0.2.3, lemmatized word forms are submitted to the
`PatternAnalyzer`, increasing the accuracy of polarity values. New in
version 0.2.7: return type of `.sentiment` is now adapted to the main
[TextBlob](http://textblob.readthedocs.org/en/dev/) library
(`:rtype: namedtuple`).

``` python
>>> blob.words.lemmatize()
WordList(['das', 'sein', 'ein', 'hässlich', 'Auto'])
>>> from textblob_de.lemmatizers import PatternParserLemmatizer
>>> _lemmatizer = PatternParserLemmatizer()
>>> _lemmatizer.lemmatize("Das ist ein hässliches Auto.")
[('das', 'DT'), ('sein', 'VB'), ('ein', 'DT'), ('hässlich', 'JJ'), ('Auto', 'NN')]
```

Access to `pattern` API in Python3
----------------------------------

``` python
>>> from textblob_de.packages import pattern_de as pd
>>> print(pd.attributive("neugierig", gender=pd.FEMALE, role=pd.INDIRECT, article="die"))
neugierigen
```

Note

Alternatively, the path to `textblob_de/ext` can be added to the
`PYTHONPATH`, which allows the use of `pattern.de` in almost the same
way as described in its
[Documentation](http://www.clips.ua.ac.be/pages/pattern-de). The only
difference is that you will have to prepend an underscore:
`from _pattern.de import ...`. This is a precautionary measure in case
the `pattern` library gets native Python3 support in the future.

Documentation and API Reference
-------------------------------

-   <http://textblob-de.readthedocs.org/en/latest>

Requirements
------------

-   Python &gt;= 3.6

TODO
----

-   [Planned
    Extensions](http://textblob-de.readthedocs.org/en/latest/extensions.html)
-   Additional PoS tagging options, e.g. NLTK tagging (`NLTKTagger`)
-   Improve noun phrase extraction (e.g. based on `RFTagger` output)
-   Improve sentiment analysis (find suitable subjectivity scores)
-   Improve functionality of `Sentence()` and `Word()` objects
-   Adapt more tests from the main
    [TextBlob](http://textblob.readthedocs.org/en/dev/) library (esp.
    for `TextBlobDE()` in `test_blob.py`)

License
-------

MIT licensed. See the bundled
[LICENSE](https://github.com/markuskiller/textblob-de/blob/master/LICENSE)
file for more details.

Thanks
------

Coded with Wing IDE (free open source developer license)

[![Python IDE for Python - wingware.com](https://wingware.com/images/wingware-logo-180x58.png)](https://wingware.com/store/free)


HISTORY.rst
Changelog
---------

0.4.4 (unreleased)
++++++++++++++++++

* TODO: fix docs build on RTD
* adapted setup.py to reflect only python versions on which the tests are run
* Updated tox/Travis-CI config files to test only on python3 versions that are still supported (>=3.6) and on python2.7 / pypy for legacy support
* Merged `Pull Request #21 <https://github.com/markuskiller/textblob-de/pull/21>`_ (thanks @anetschka)
* Fixed tests
* Extended list of abbreviations


0.4.3 (03/01/2019)
++++++++++++++++++

* Added support for Python3.7 (``StopIteration --> return``) `Pull Request #18 <https://github.com/markuskiller/textblob-de/pull/18>`_ (thanks @andrewmfiorillo)
* Fixed tests for Google translation examples
* Updated tox/Travis-CI config files to include latest Python & pypy versions
* Updated sphinx_rtd_theme to version 0.4.2 to fix rendering problems on `RTD <http://textblob-de.readthedocs.org>`_
* Updated ``setup.py publish`` commands, ``Makefile`` & ``Manifest.in`` to new PyPI (using ``twine``)

0.4.2 (02/05/2015)
++++++++++++++++++

* Removed dependency on `NLTK, <https://github.com/nltk/nltk/>`_ as it already is a `TextBlob <http://textblob.readthedocs.org/en/dev/>`_ dependency
* Temporary workaround for `NLTK Issue #824 <https://github.com/nltk/nltk/issues/824>`_ for tox/Travis-CI
* (update 13/01/2015) `NLTK Issue #824 <https://github.com/nltk/nltk/issues/824>`_ fixed, workaround removed
* Enabled ``pattern`` tagset conversion (``'penn'|'universal'|'stts'``) for ``PatternTagger``
* Added tests for tagset conversion
* Fixed test for Arabic translation example (Google translation has changed)
* Added tests for lemmatizer
* Bugfix: ``PatternAnalyzer`` no longer breaks on subsequent ocurrences of the same ``(word, tag)`` pairs on Python3 see comments to `Pull Request #11 <https://github.com/markuskiller/textblob-de/pull/11>`_
* Bugfix/performance enhancement: Sentiment dictionary in ``PatternAnalyzer`` no longer reloaded for every sentence `Pull Request #11 <https://github.com/markuskiller/textblob-de/pull/11>`_ (thanks @Arttii)

0.4.1 (03/10/2014)
++++++++++++++++++

* Docs hosted on `RTD <http://textblob-de.readthedocs.org>`_
* Removed dependency on nltk's depricated ``PunktWordTokenizer`` and replaced it with ``TreebankWordTokenizer`` see  `nltk/nltk#746 (comment) <https://github.com/nltk/nltk/pull/746#issuecomment-57625756>`_ for details

0.4.0 (17/09/2014)
++++++++++++++++++

* Fixed `Issue #7 <https://github.com/markuskiller/textblob-de/issues/7>`_ (restore ``textblob>=0.9.0`` compatibility)
* Depend on ``nltk3``. Vendorized ``nltk`` was removed in ``textblob>=0.9.0``
* Fixed ``ImportError`` on Python2 (``unicodecsv``)


0.3.1 (29/08/2014)
++++++++++++++++++

* Improved ``PatternParserNPExtractor`` (less false positives in verb filter)
* Made sure that all keyword arguments with default ``None`` are checked with ``is not None``
* Fixed shortcut to ``_pattern.de`` in vendorized library
* Added ``Makefile`` to facilitate development process
* Added docs and API reference

0.3.0 (14/08/2014)
++++++++++++++++++

* Fixed `Issue #5 <https://github.com/markuskiller/textblob-de/issues/5>`_ (text + space + period)

0.2.9 (14/08/2014)
++++++++++++++++++

* Fixed tokenization in ``PatternParser`` (if initialized manually, punctuation was not always separated from words)
* Improved handling of empty strings (Issue #3) and of strings containing single punctuation marks (Issue #4) in ``PatternTagger`` and ``PatternParser``
* Added tests for empty strings and for strings containing single punctuation marks

0.2.8 (14/08/2014)
++++++++++++++++++

* Fixed `Issue #3 <https://github.com/markuskiller/textblob-de/issues/3>`_ (empty string)
* Fixed `Issue #4 <https://github.com/markuskiller/textblob-de/issues/4>`_ (space + punctuation)

0.2.7 (13/08/2014)
++++++++++++++++++

* Fixed `Issue #1 <https://github.com/markuskiller/textblob-de/issues/1>`_ lemmatization of strings containing a forward slash (``/``)
* Enhancement `Issue #2 <https://github.com/markuskiller/textblob-de/issues/2>`_ use the same rtype as ``textblob`` for sentiment detection.
* Fixed tokenization in ``PatternParserLemmatizer``

0.2.6 (04/08/2014)
++++++++++++++++++

* Fixed ``MANIFEST.in`` for package data in ``sdist``

0.2.5 (04/08/2014)
++++++++++++++++++

* ``sdist`` is non-functional as important files are missing due to a misconfiguration in ``MANIFEST.in`` - does not affect ``wheels``
* Major internal refactoring (but no backwards-incompatible API changes) with the aim of restoring complete compatibility to original ``pattern>=2.6`` library on Python2
* Separation of ``textblob`` and ``pattern`` code
* On Python2 the vendorized version of ``pattern.text.de`` is only used if original is not installed (same as ``nltk``)
* Made ``pattern.de.pprint`` function and all parser keywords accessible to customise parser output
* Access to complete ``pattern.text.de`` API on Python2 and Python3 ``from textblob_de.packages import pattern_de as pd``
* ``tox`` passed on all major platforms (Win/Linux/OSX)

0.2.3 (26/07/2014)
++++++++++++++++++

* Lemmatizer: ``PatternParserLemmatizer()`` extracts lemmata from Parser output
* Improved polarity analysis through look-up of lemmatised word forms

0.2.2 (22/07/2014)
++++++++++++++++++

* Option: Include punctuation in ``tags``/``pos_tags`` properties (``b = TextBlobDE(text, tagger=PatternTagger(include_punc=True))``)
* Added ``BlobberDE()`` class initialized with German models
* ``TextBlobDE()``, ``Sentence()``, ``WordList()`` and ``Word()`` classes are now all initialized with German models
* Restored complete API compatibility with ``textblob.tokenizers`` module of the main `TextBlob <http://textblob.readthedocs.org/en/dev/>`_ library

0.2.1 (20/07/2014)
++++++++++++++++++

* Noun Phrase Extraction: ``PatternParserNPExtractor()`` extracts NPs from Parser output
* Refactored the way ``TextBlobDE()`` passes on arguments and keyword arguments to individual tools

textblob_de/compat.py
# -*- coding: utf-8 -*-
# Code adapted from the main `TextBlob`_ library.
#
# :repo: `https://github.com/sloria/TextBlob`_
# :source: textblob/compat.py
# :version: 2013-12-16 (e7f9cf9)
#
# :modified: 2014-09-17 <m.killer@langui.ch>
#
'''Compatibility module for shared code base on Python2 and Python3.
'''
import os
import sys


PY2 = int(sys.version[0]) == 2
PY26 = PY2 and int(sys.version_info[1]) < 7

if PY2:
    from itertools import imap, izip
    import urllib2 as request
    from urllib import quote as urlquote
    from urllib import urlencode
    _FileNotFoundError = IOError
    # ``codecs.open`` with encoding keyword on Python2
    from codecs import open as _open
    _which = "compat"
    text_type = unicode
    binary_type = str
    string_types = (str, unicode)
    unicode = unicode
    basestring = basestring
    imap = imap
    izip = izip
    # not quite sure why this ImportError did not
    # seem to occur earlier ...
    try:
        import unicodecsv as csv
    except ImportError:
        import csv

    def implements_to_string(cls):
        '''Class decorator that renames __str__ to __unicode__ and
        modifies __str__ that returns utf-8.
        '''
        cls.__unicode__ = cls.__str__
        cls.__str__ = lambda x: x.__unicode__().encode('utf-8')
        return cls
else:  # PY3
    from urllib import request
    from urllib.parse import quote as urlquote
    from urllib.parse import urlencode
    # ``io.open`` with encoding keyword on Python3
    from io import open as _open
    try:
        from shutil import which as _which
    except ImportError:
        _which = "compat"
    # add pypy3 compatibilty
    try:
        _FileNotFoundError = FileNotFoundError
    except NameError:
        _FileNotFoundError = IOError
    text_type = str
    binary_type = bytes
    string_types = (str,)
    unicode = str
    basestring = (str, bytes)
    imap = map
    izip = zip
    import csv

    implements_to_string = lambda x: x


def with_metaclass(meta, *bases):
    """Defines a metaclass.

    Creates a dummy class with a dummy metaclass. When subclassed, the
    dummy metaclass is used, which has a constructor that instantiates a
    new class from the original parent. This ensures that the dummy
    class and dummy metaclass are not in the inheritance tree.

    Credit to Armin Ronacher.

    """
    class metaclass(meta):
        __call__ = type.__call__
        __init__ = type.__init__

        def __new__(cls, name, this_bases, d):
            if this_bases is None:
                return type.__new__(cls, name, (), d)
            return meta(name, bases, d)
    return metaclass('temporary_class', None, {})


#--- STRING FUNCTIONS --------------------------------------------------

# Source: https://github.com/clips/pattern/pattern/text/__init__.py
# git-commit: 2014-05-10 (2f944cb)

# Modified: 2014-08-04 Markus Killer <m.killer@langui.ch>

# Latin-1 (ISO-8859-1) encoding is identical to Windows-1252 except for
# the code points 128-159:
# Latin-1 assigns control codes in this range, Windows-1252 has
# characters, punctuation, symbols
# assigned to these code points.

# String functions
def decode_string(v, encoding="utf-8"):
    """Returns the given value as a Unicode string (if possible)."""
    if isinstance(encoding, basestring):
        encoding = ((encoding,),) + (("windows-1252",), ("utf-8", "ignore"))
    if isinstance(v, binary_type):
        for e in encoding:
            try:
                return v.decode(*e)
            except:
                pass
        return v
    return unicode(v)


def encode_string(v, encoding="utf-8"):
    """Returns the given value as a Python byte string (if possible)."""
    if isinstance(encoding, basestring):
        encoding = ((encoding,),) + (("windows-1252",), ("utf-8", "ignore"))
    if isinstance(v, unicode):
        for e in encoding:
            try:
                return v.encode(*e)
            except:
                pass
        return v
    return str(v)

decode_utf8 = decode_string
encode_utf8 = encode_string


def get_external_executable(_exec):
    _exec_path = _which(_exec)
    if _exec_path:
        return _exec_path
    else:
        sys.exit("Required executable '{}' not found on this system. "
                 "Please install or add to PATH".format(_exec))

# Code adapted from Python 3.4 standard library ``shutil.py`` (lines: 1067-1127)
# to work on Python2
# Source: http://stackoverflow.com/questions/9877462 [accessed: 27/08/2014]


def _shutil_which(cmd, mode=os.F_OK | os.X_OK, path=None):
    """Given a command, mode, and a PATH string, return the path which conforms
    to the given mode on the PATH, or None if there is no such file.

    `mode` defaults to os.F_OK | os.X_OK. `path` defaults to the result
    of os.environ.get("PATH"), or can be overridden with a custom search
    path.

    """
    # Check that a given file can be accessed with the correct mode.
    # Additionally check that `file` is not a directory, as on Windows
    # directories pass the os.access check.
    def _access_check(fn, mode):
        return (os.path.exists(fn) and os.access(fn, mode)
                and not os.path.isdir(fn))

    # If we're given a path with a directory part, look it up directly rather
    # than referring to PATH directories. This includes checking relative to the
    # current directory, e.g. ./script
    if os.path.dirname(cmd):
        if _access_check(cmd, mode):
            return cmd
        return None

    if path is None:
        path = os.environ.get("PATH", os.defpath)
    if not path:
        return None
    path = path.split(os.pathsep)

    if sys.platform == "win32":
        # The current directory takes precedence on Windows.
        if not os.curdir in path:
            path.insert(0, os.curdir)

        # PATHEXT is necessary to check on Windows.
        pathext = os.environ.get("PATHEXT", "").split(os.pathsep)
        # See if the given file matches any of the expected path extensions.
        # This will allow us to short circuit when given "python.exe".
        # If it does match, only test that one, otherwise we have to try
        # others.
        if any([cmd.lower().endswith(ext.lower()) for ext in pathext]):
            files = [cmd]
        else:
            files = [cmd + ext for ext in pathext]
    else:
        # On other platforms you don't have things like PATHEXT to tell you
        # what file suffixes are executable, so just pass on cmd as-is.
        files = [cmd]

    seen = set()
    for dir in path:
        normdir = os.path.normcase(dir)
        if normdir not in seen:
            seen.add(normdir)
            for thefile in files:
                name = os.path.join(dir, thefile)
                if _access_check(name, mode):
                    return name
    return None

if _which == "compat":
    _which = _shutil_which
